app/src/main/java/com/github/factotum_sdp/factotum/ui/roadbook/DRecordList.kt

```diff
@@ -0,0 +1,90 @@
+package com.github.factotum_sdp.factotum.ui.roadbook
+
+import com.github.factotum_sdp.factotum.data.DestinationRecord
+
+
+private fun displayedRecords(allRecords: List<DestinationRecord>,
+                             archived: List<DestinationRecord>,
+                             showArchived: Boolean): List<DestinationRecord> {
+    if (showArchived)
+        return allRecords
+    return allRecords.minus(archived.toSet())
+}
+
+/**
+ *   Immutable Decorator for a List<DestinationRecord>
+ *
+ * - Allows archived records management through a private field containing all the archived items
+ * - Allows choice at construction time for the main List, whether it contains or not the archived records.
+ */
+class DRecordList(private val allRecords: List<DestinationRecord> = listOf(),
+                  private val archived: List<DestinationRecord> = listOf(),
+                  private val showArchived: Boolean = false)
+    : List<DestinationRecord> by displayedRecords(allRecords, archived, showArchived) {
+
+    private val archivedSet = archived.toSet() // For performance
+
+    /**
+     * Check if the record at position "index" is archived
+     *
+     * @param index: Int index of the target DestinationRecord
+     * @return Boolean whether the specified record is archived or not
+     */
+    fun isArchived(index: Int): Boolean {
+        return archivedSet.contains(this[index])
+    }
+
+    /**
+     * Update the current main List (this) in an immutable way,
+     * keeping all the others current settings.
+     *
+     * @param ls: List<DestinationRecord> The new list to replace the current (this) main List
+     * @return a new DRecordList with the main List updated
+     */
+    fun newDisplayedList(ls: List<DestinationRecord>): DRecordList {
+        val allRecs =
+            if(showArchived)
+                ls
+            else
+                archived.plus(ls)
+        return DRecordList(allRecs, archived, showArchived)
+    }
+
+    /**
+     * Archive a record
+     *
+     * @param index: Int index of the target DestinationRecord
+     * @return a new DRecordList with the specified record added to the archived records
+     */
+    fun archiveRecord(index: Int): DRecordList {
```

Comment:
Same here `withRecordArchived` coult be a better name

Reply:
I prefer not as it does not affect the displayedList (i.e the main list of the class directly).

When calling that function you just want to archive a particular record, nothing more.

Then if you want to modify the main list of the DRecordList, the one decorated.
You call the dRecordList.withArchived() which affects the list.